Add integration with pgsql in vagga

DrEdi opened the pull request:
Add new app where we create admin for db
Integrate vagga with pgsql
Refresh requirements

DrEdi committed:
Add integration with pgsql in vagga

Add new app where we create admin for db
Integrate vagga with pgsql
Refresh reirements

Miseracle commented on common/migrations/0001_create_superuser.py:
```diff
@@ -0,0 +1,26 @@
+
+
+def create_superuser(apps, schema_editor):
+    """Module is used when we start container with PgSQL."""
```
Is it a docstring for a module?

webknjaz commented on common/__init__.py:
```diff
@@ -0,0 +1 @@
+"""Empty file."""
```
Why do you need this file?

webknjaz replied:
Current docstring does not make sense, but you can make it useful by adding a docstring describing the whole `common` package

webknjaz commented on common/migrations/0001_create_superuser.py:
```diff
@@ -0,0 +1,26 @@
+    """Module is used when we start container with PgSQL."""
+    User = apps.get_model('auth', 'User')
+    User.objects.create(username='admin',
+                        email='admin@example.com',
```
I'd read this from environment with a default fallback. Like `os.getenv('ADMIN_EMAIL', 'admin@example.com')`

webknjaz replied:
(or define it via `django-environ` in settings)

webknjaz commented on common/migrations/0001_create_superuser.py:
```diff
@@ -0,0 +1,26 @@
+    User = apps.get_model('auth', 'User')
+    User.objects.create(username='admin',
+                        email='admin@example.com',
+                        password=make_password('12345678'),
```
I would generate random password

DrEdi replied:
But how will I know this random generated password?
From environment?

webknjaz commented on common/migrations/0001_create_superuser.py:
```diff
@@ -0,0 +1,26 @@
+
+
+class Migration(migrations.Migration):
+    """Config class."""
```
I think it's a migration plan for superuser creation.

webknjaz commented on common/migrations/__init__.py:
```diff
@@ -0,0 +1 @@
+"""Empty file."""
```
Why did you create this file?

DrEdi replied:
Auto created
Would it better to remove?

webknjaz replied:
As u wish, but remove that stupid docstring

webknjaz commented on vagga.yaml:
```diff
@@ -2,19 +2,41 @@ containers:
     setup:
       - !Alpine v3.4
+      - !BuildDeps
+        - postgresql-dev
```
Why it's not `libpq` as in `django` container?

DrEdi replied:
We need it in runtime

webknjaz replied:
so why it's different here?

DrEdi replied:
Documentations says that postgresql needed for building psycopg, libpq - needed for it in runtime.
There's no other info.
But I think difference is also in !BuildDeps and !Install

webknjaz replied:
`!BuildDeps` are removed after container is built. `!Install` packages are available in runtime.
But the question here is that you install `psycopg2` python package and it requires some package during build time which you don't provide in `django` container. This means that `libpq` is enough for build, so why not use the same package here?

webknjaz commented on vagga.yaml:
```diff
@@ -2,19 +2,41 @@ containers:
       - !Py3Requirements requirements.txt
+  postgres:
+    setup:
+    - !Ubuntu trusty
```
JFYI there's newer Ubuntu version, which is `xenial`

webknjaz replied:
P.S. You could also base it on Alpine

webknjaz commented on vagga.yaml:
```diff
@@ -24,3 +46,26 @@ commands:
+    children:
+      app: !Command
+        container: django
+        environ:
```
Nit: you could add this right into container description, since this is not the only command needing this var

webknjaz commented on vagga.yaml:
```diff
@@ -24,3 +46,26 @@ commands:
+        environ:
+          DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/django
+        run: |
+            touch /work/.dbcreation # Create lock file
```
I would follow python's conventions here as well and use 2 whitespaces before `#`

webknjaz commented on vagga.yaml:
```diff
@@ -24,3 +46,26 @@ commands:
+        container: postgres
+        run: |
+            chown postgres:postgres $PGDATA;
+            su postgres -c "$PG_BIN/pg_ctl initdb";
```
I think, DB creation should not repeat on every command run, you need a separate command for initializing the db

webknjaz replied:
Look for init command in volume definition

webknjaz commented on vagga.yaml:
```diff
@@ -2,19 +2,41 @@ containers:
+      PG_PASSWORD: vagga
+      PG_BIN: /usr/lib/postgresql/9.3/bin
+    volumes:
+      /data: !Tmpfs
```
Use persistent named volume

DrEdi replied:
Opps
`vagga.yaml:24:5: Validation Error: Keys {"volume"} are not expected`

webknjaz commented on common/migrations/0001_create_superuser.py:
```diff
@@ -0,0 +1,29 @@
+
+
+def create_superuser(apps, schema_editor):
+    """Function is used when we start container with PgSQL."""
```
Write in imperative style:
```python
"""Creates superuser for the app at the first start"""
```

webknjaz commented on common/migrations/0001_create_superuser.py:
```diff
@@ -0,0 +1,29 @@
+from django.contrib.auth.hashers import make_password
+import environ
+
+env = environ.Env(DEBUG=(bool, False),)
```
1) Please don't forget reading `.env` into environment.
2) `DEBUG` var is not needed here.
3) It is better to move declaration of `ADMIN_EMAIL` into settings and just import it from there. Keeping all related things in one place helps a lot :)

webknjaz commented on vagga.yaml:
```diff
@@ -2,20 +2,65 @@ containers:
+  postgres:
+    setup:
+    - !Ubuntu xenial
+    - !EnsureDir /data
```
According to doc for `!Persistent`, `!EnsureDir` is not needed for such mount point. (Please check this)

DrEdi replied:
I have an error without this 
`No such file or directory (os error 2) (source: exists, target: missing, mapped-root)`

webknjaz commented on vagga.yaml:
```diff
@@ -24,3 +69,18 @@ commands:
+        container: postgres
+        user-id: 200
+        group-id: 200
+        run: exec $PG_BIN/postgres -F --port=$PG_PORT
```
to use EXEC syntax in vagga you may just pass an array. like:
```yaml
run: [exec, $PG_BIN/postgres, -F, --port=$PG_PORT]
```

DrEdi replied:
This also come with error 
`ERROR:vagga::wrapper: Error executing run-postgres: Command exec not found in "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
---------- Process db:8849 exited with code 124. Shutting down -----------`

webknjaz replied:
my bad, remove `[exec]` from the sequence

DrEdi replied:
Something gone wrong
`ERROR:vagga::wrapper: Error executing run-postgres: Error running <Command "$PG_BIN/postgres" "-F" "--port=$PG_PORT";`

webknjaz replied:
It looks like env vars are not getting interpolated in exec, so leave it as is for now.

webknjaz reviewed (changes requested):
I've pointed out some suggestions, but I'll need to return and check the logic of PG commands

webknjaz commented on common/migrations/0001_create_superuser.py:
```diff
@@ -3,14 +3,15 @@
     User = apps.get_model('auth', 'User')
     User.objects.create(username='admin',
-                        email=env('ADMIN_EMAIL', 'admin@example.com'),
+                        email=env('ADMIN_EMAIL'),
```
why did you remove default value?

webknjaz commented on common/migrations/0001_create_superuser.py:
```diff
@@ -3,14 +3,15 @@
 from django.contrib.auth.hashers import make_password
 import environ
```
don't use `environ` in this module

webknjaz commented on common/migrations/0001_create_superuser.py:
```diff
@@ -0,0 +1,27 @@
+"""Module for creating DB-admin."""
+from django.db import migrations
+from django.contrib.auth.hashers import make_password
+from blog.settings import ADMIN_EMAIL
```
https://docs.djangoproject.com/en/1.10/topics/settings/#using-settings-in-python-code

webknjaz commented on vagga.yaml:
```diff
@@ -2,20 +2,65 @@ containers:
+    environ:
+      PGDATA: /data
+      PG_PORT: 5433
+      PG_DB: test
```
Don't use test DB, create your own.

DrEdi replied:
ok

webknjaz commented on vagga.yaml:
```diff
@@ -24,3 +69,18 @@ commands:
+        environ:
+          DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/test
+        run: |
+            python3 manage.py migrate
```
You should normally wait until PG becomes available to be able to apply migrations.

webknjaz commented on common/migrations/__init__.py:
```diff
@@ -0,0 +1 @@
+"""This file is needed for migration admin to DB."""
```
Describe the migrations package instead.

DrEdi replied:
done

webknjaz commented on vagga.yaml:
```diff
@@ -75,7 +75,7 @@ commands:
         container: django
         environ:
-          DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/test
+          DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/mainDB
```
Why didn't you move it to container declaration?

DrEdi replied:
It will not work if I move it to container:
`Is the server running on host "127.0.0.1" and accepting
	TCP/IP connections on port 8458?
`

webknjaz commented on blog/settings.py:
```diff
@@ -111,17 +110,8 @@
 
-public_root = root.path('public/')
-
+public_root = env.path('STATIC_ROOT', BASE_DIR('static'))
```
I'm not sure whether this is correct

DrEdi replied:
It works. I've tested it before push

webknjaz replied:
So what does this invocation mean?

webknjaz reviewed (changes requested):
Please read my old comments, which are collapsed by now as well

DrEdi committed:
Describe migration package in common

DrEdi committed:
Divide commands in vagga

webknjaz commented on vagga.yaml:
```diff
@@ -63,24 +63,22 @@ commands:
             python3 manage.py migrate
             python3 manage.py runserver --insecure
-      db: !Command
+  db: &run-db !Command
```
run-db plz

DrEdi replied:
done

DrEdi committed:
Fix command name

webknjaz commented on common/migrations/__init__.py:
```diff
@@ -0,0 +1,7 @@
+"""
```
Mind your grammar.

webknjaz commented on blog/settings.py:
```diff
@@ -3,11 +3,10 @@
 
+# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
+BASE_DIR = environ.Path(__file__) - 2
+env.read_env(str(BASE_DIR.path('.env')))
```
Have you tried doing it [this way](https://github.com/open-craft-guild/crm-smsfly/blob/master/SMSFlyCRM/settings.py#L17)?

DrEdi replied:
Yes. I have the same problem as before
`UserWarning: /work/blog/.env doesn't exist - if you're not configuring your environment separately, create one.
  "environment separately, create one." % env_file)`

webknjaz replied:
Have u tried leaving `environ.Env.read_env()`?

DrEdi replied:
Yep. I always have a warning like in previous comment.

webknjaz replied:
why ` - 2`?

DrEdi replied:
I tested it ' by trial and error'. This is the only one thing where it sees .env file.
Also in stackoverfow where I found how to fix it was this variant.

webknjaz commented on vagga.yaml:
```diff
@@ -2,25 +2,83 @@ containers:
+      set -ex
+      ls -la /data
+      $PG_BIN/pg_ctl initdb
+      $PG_BIN/pg_ctl -w -o '-F --port=$PG_PORT -k /tmp' start
```
I `/tmp` should be `/data`. Does it preserve DB across restarts?

DrEdi replied:
Yes, everything is working ok

webknjaz replied:
Please clean volumes via `vagga _clean --volumes` and check again.

DrEdi replied:
I've got fully clean project. But after add and reboot - new post still there, it's ok?

webknjaz replied:
yes

webknjaz replied:
Still change `/tmp` to `$PGDATA`

webknjaz commented on vagga.yaml:
```diff
@@ -2,25 +2,83 @@ containers:
+        owner-uid: 200
+        owner-gid: 200
+        init-command: _pg-init
+      /run: !Tmpfs
```
Why did you add this?

DrEdi replied:
It is in vagga documentation. I thought it'd be better not to change anything

webknjaz replied:
It seems to be working fine w/o it

webknjaz commented on vagga.yaml:
```diff
@@ -2,25 +2,83 @@ containers:
+        container: postgres
+        user-id: 200
+        group-id: 200
+        run: exec $PG_BIN/postgres -F --port=$PG_PORT
```
Why don't you pass `-k ...` to this command?

DrEdi replied:
Answered in gitter :)

webknjaz replied:
Please add `-k $PGDATA` just in case

webknjaz commented on vagga.yaml:
```diff
@@ -2,25 +2,83 @@ containers:
+      - !Install
+        - libpq
       - !Py3Requirements requirements.txt
+    environ:
```
I would also add `environment-file` directive to this container declaration.

DrEdi replied:
Ok.

webknjaz commented on common/migrations/__init__.py:
```diff
@@ -0,0 +1,7 @@
+"""
+Package for creating admin account.
```
Fixture for admin account.

webknjaz commented on common/migrations/__init__.py:
```diff
@@ -0,0 +1,7 @@
+"""
+Package for creating admin account.
+
+When you run project with pgsqlDB, admin model
```
```
This will automatically add pre-configured admin account
to your db during migration. Specify admin's email via
ADMIN_EMAIL environment variable. The password is
hardcoded to 12345678, don't forget to reset it.
```

DrEdi replied:
Changed.
Thanks

DrEdi committed:
Change description of package

DrEdi committed:
Add environ-file to container declaration

webknjaz commented on vagga.yaml:
```diff
@@ -21,7 +21,7 @@ containers:
         - libpq
       - !Py3Requirements requirements.txt
     environ:
-          DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/mainDB
```
Why did you remove this var? It suits its purpose well.

webknjaz commented on vagga.yaml:
```diff
@@ -21,7 +21,7 @@ containers:
       - !Py3Requirements requirements.txt
     environ:
-          DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/mainDB
+      environ-file: /work/blog/.env
```
the indentation is broken here

DrEdi replied:
You mean it must be under 'environ' without tab, right?

DrEdi committed:
Fix spaces, return DB variable

webknjaz reviewed (commented):
Could you please also specify the minimum version of `vagga` needed for this config to work?

DrEdi committed:
Fix vagga

Change dir in run command
Add additional info to run-db command
Add anchor to django container

DrEdi committed:
Specify minimum version of vagga

webknjaz commented on vagga.yaml:
```diff
@@ -1,26 +1,87 @@
+minimum-vagga: v0.7.0-1-gc613573
```
I don't think vagga is able to compare commit hash part of the version, please cut it off.

DrEdi replied:
Oh, I thought it's required.

DrEdi committed:
Fix vagga minimum version

webknjaz commented on vagga.yaml:
```diff
@@ -1,26 +1,87 @@
+minimum-vagga: v0.7.0-1
```
`v0.7.0`is enough
